intel-media-driver fails on 32bits

### Describe the bug

The build of `intel-media-driver` fails on i686.

### Steps To Reproduce

```
nix-build -p pkgsi686Linux.intel-media-driver
```

Fails with:

```console
warning: unknown setting 'experimental-features'
these derivations will be built:
  /nix/store/jynx74kc1n7q0wfsmzk3wzsrkzsc0w93-intel-media-driver-21.3.3.drv
  /nix/store/5631yxw34z6phkqkrh34mhb0c4clgwp8-shell.drv
warning: unknown setting 'experimental-features'
building '/nix/store/jynx74kc1n7q0wfsmzk3wzsrkzsc0w93-intel-media-driver-21.3.3.drv'...
unpacking sources
unpacking source archive /nix/store/3j118ymdkv9xnpva1bnj30mdfpqbqsyn-source
source root is source
patching sources
configuring
fixing cmake files...
cmake flags: -DCMAKE_FIND_USE_SYSTEM_PACKAGE_REGISTRY=OFF -DCMAKE_FIND_USE_PACKAGE_REGISTRY=OFF -DCMAKE_EXPORT_NO_PACKAGE_REGISTRY=ON -DCMAKE_BUILD_TYPE=Release -DCMAKE_SKIP_BUILD_RPATH=ON -DBUILD_TESTING=OFF -DCMAKE_INSTALL_LOCALEDIR=/nix/store/qrwm3c7q0ssg0jai0d1n098b9dz1flaq-intel-media-driver-21.3.3/share/locale -DCMAKE_INSTALL_LIBEXECDIR=/nix/store/qrwm3c7q0ssg0jai0d1n098b9dz1flaq-intel-media-driver-21.3.3/libexec -DCMAKE_INSTALL_LIBDIR=/nix/store/qrwm3c7q0ssg0jai0d1n098b9dz1flaq-intel-media-driver-21.3.3/lib -DCMAKE_INSTALL_DOCDIR=/nix/store/qrwm3c7q0ssg0jai0d1n098b9dz1flaq-intel-media-driver-21.3.3/share/doc/IntelMediaDriver -DCMAKE_INSTALL_INFODIR=/nix/store/qrwm3c7q0ssg0jai0d1n098b9dz1flaq-intel-media-driver-21.3.3/share/info -DCMAKE_INSTALL_MANDIR=/nix/store/qrwm3c7q0ssg0jai0d1n098b9dz1flaq-intel-media-driver-21.3.3/share/man -DCMAKE_INSTALL_OLDINCLUDEDIR=/nix/store/rf9mkpphsfbmc7qj3zmxbky55jpdcgcy-intel-media-driver-21.3.3-dev/include -DCMAKE_INSTALL_INCLUDEDIR=/nix/store/rf9mkpphsfbmc7qj3zmxbky55jpdcgcy-intel-media-driver-21.3.3-dev/include -DCMAKE_INSTALL_SBINDIR=/nix/store/qrwm3c7q0ssg0jai0d1n098b9dz1flaq-intel-media-driver-21.3.3/sbin -DCMAKE_INSTALL_BINDIR=/nix/store/qrwm3c7q0ssg0jai0d1n098b9dz1flaq-intel-media-driver-21.3.3/bin -DCMAKE_INSTALL_NAME_DIR=/nix/store/qrwm3c7q0ssg0jai0d1n098b9dz1flaq-intel-media-driver-21.3.3/lib -DCMAKE_POLICY_DEFAULT_CMP0025=NEW -DCMAKE_OSX_SYSROOT= -DCMAKE_FIND_FRAMEWORK=LAST -DCMAKE_STRIP=/nix/store/pndi2rnmqfhk4cajblsrnml2fx8w2mz0-binutils-2.35.1/bin/strip -DCMAKE_RANLIB=/nix/store/pndi2rnmqfhk4cajblsrnml2fx8w2mz0-binutils-2.35.1/bin/ranlib -DCMAKE_AR=/nix/store/pndi2rnmqfhk4cajblsrnml2fx8w2mz0-binutils-2.35.1/bin/ar -DCMAKE_C_COMPILER=gcc -DCMAKE_CXX_COMPILER=g++ -DCMAKE_INSTALL_PREFIX=/nix/store/qrwm3c7q0ssg0jai0d1n098b9dz1flaq-intel-media-driver-21.3.3 -DINSTALL_DRIVER_SYSCONF=OFF -DLIBVA_DRIVERS_PATH=/nix/store/qrwm3c7q0ssg0jai0d1n098b9dz1flaq-intel-media-driver-21.3.3/lib/dri -DMEDIA_RUN_TEST_SUITE=OFF
-- The C compiler identification is GNU 10.3.0
-- The CXX compiler identification is GNU 10.3.0
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /nix/store/mpmy3726h9r98mzjl6pddhxhkgccspb9-gcc-wrapper-10.3.0/bin/gcc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /nix/store/mpmy3726h9r98mzjl6pddhxhkgccspb9-gcc-wrapper-10.3.0/bin/g++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- CMAKE_INSTALL_PREFIX = /nix/store/qrwm3c7q0ssg0jai0d1n098b9dz1flaq-intel-media-driver-21.3.3
CMake Deprecation Warning at cmrtlib/CMakeLists.txt:22 (cmake_minimum_required):
  Compatibility with CMake < 2.8.12 will be removed from a future version of
  CMake.

  Update the VERSION argument <min> value or use a ...<max> suffix to tell
  CMake that the project does not need compatibility with older versions.


CMake Deprecation Warning at cmrtlib/linux/CMakeLists.txt:21 (cmake_minimum_required):
  Compatibility with CMake < 2.8.12 will be removed from a future version of
  CMake.

  Update the VERSION argument <min> value or use a ...<max> suffix to tell
  CMake that the project does not need compatibility with older versions.


-- Found PkgConfig: pkg-config (found version "0.29.2")
-- Checking for module 'libva>=1.8.0'
--   Found libva, version 1.12.0
*BUILD_TYPE not defined, default to: release
-- Found X11: /nix/store/mix0g3agk4a10j76vq84ixzakmjxjlz3-xorgproto-2021.4/include
-- Looking for XOpenDisplay in /nix/store/yqzbpjj7mwsbxnjdxpyx9zc1mpay42zh-libX11-1.7.2/lib/libX11.so
-- Looking for XOpenDisplay in /nix/store/yqzbpjj7mwsbxnjdxpyx9zc1mpay42zh-libX11-1.7.2/lib/libX11.so - found
-- Looking for gethostbyname
-- Looking for gethostbyname - found
-- Looking for connect
-- Looking for connect - found
-- Looking for remove
-- Looking for remove - found
-- Looking for shmat
-- Looking for shmat - found
-- Checking for module 'igdgmm'
--   Found igdgmm, version 11.3.0
-- media -- PLATFORM = linux
-- media -- ARCH = 64
-- media -- CMAKE_CURRENT_LIST_DIR = /build/source/media_driver
-- media -- INCLUDED_LIBS = 
-- media -- LIB_NAME = iHD_drv_video
-- media -- OUTPUT_NAME = 
-- media -- BUILD_TYPE/UFO_BUILD_TYPE/CMAKE_BUILD_TYPE = release/release/Release
-- media -- LIBVA_INSTALL_PATH = 
-- media -- MEDIA_VERSION = 21.3.3
-- media -- X11 Found
-- LIBVA_DRIVERS_PATH = /nix/store/qrwm3c7q0ssg0jai0d1n098b9dz1flaq-intel-media-driver-21.3.3/lib/dri
-- Configuring done
-- Generating done
CMake Warning:
  Manually-specified variables were not used by the project:

    BUILD_TESTING
    CMAKE_EXPORT_NO_PACKAGE_REGISTRY


-- Build files have been written to: /build/source/build
cmake: enabled parallel building
building
build flags: -j8 -l8 SHELL=/nix/store/ra7f5niavfph8mhfm2lf51nik07r1k9j-bash-5.1-p8/bin/bash
[  0%] Building CXX object cmrtlib/linux/CMakeFiles/igfxcmrt.dir/__/agnostic/share/cm_avs_state_msg_ex.cpp.o
[  0%] Building CXX object media_driver/CMakeFiles/iHD_drv_video_SSE2.dir/agnostic/common/cm/cm_mem_sse2_impl.cpp.o
[  0%] Building CXX object cmrtlib/linux/CMakeFiles/igfxcmrt.dir/__/agnostic/share/cm_rt_helpers.cpp.o
[  0%] Building CXX object media_driver/CMakeFiles/iHD_drv_video_SSE4.dir/linux/common/cm/hal/cm_mem_os_sse4_impl.cpp.o
[  0%] Building CXX object cmrtlib/linux/CMakeFiles/igfxcmrt.dir/__/agnostic/share/cm_printf_host.cpp.o
[  0%] Building CXX object cmrtlib/linux/CMakeFiles/igfxcmrt.dir/__/agnostic/hardware/cm_device.cpp.o
[  0%] Building CXX object cmrtlib/linux/CMakeFiles/igfxcmrt.dir/__/agnostic/hardware/cm_device_export.cpp.o
cc1plus: sorry, unimplemented: 64-bit mode not compiled in
make[2]: *** [media_driver/CMakeFiles/iHD_drv_video_SSE2.dir/build.make:76: media_driver/CMakeFiles/iHD_drv_video_SSE2.dir/agnostic/common/cm/cm_mem_sse2_impl.cpp.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:273: media_driver/CMakeFiles/iHD_drv_video_SSE2.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....
cc1plus: sorry, unimplemented: 64-bit mode not compiled in
make[2]: *** [media_driver/CMakeFiles/iHD_drv_video_SSE4.dir/build.make:76: media_driver/CMakeFiles/iHD_drv_video_SSE4.dir/linux/common/cm/hal/cm_mem_os_sse4_impl.cpp.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:247: media_driver/CMakeFiles/iHD_drv_video_SSE4.dir/all] Error 2
[  0%] Building CXX object media_driver/CMakeFiles/iHD_drv_video_OBJ.dir/agnostic/common/cm/cm_array.cpp.o
[  0%] Building CXX object media_driver/CMakeFiles/iHD_drv_video_OBJ.dir/agnostic/common/cm/cm_buffer_rt.cpp.o
[  0%] Building CXX object cmrtlib/linux/CMakeFiles/igfxcmrt.dir/__/agnostic/hardware/cm_surface_manager.cpp.o
[  0%] Building CXX object cmrtlib/linux/CMakeFiles/igfxcmrt.dir/__/agnostic/hardware/cm_timer.cpp.o
[  1%] Building CXX object cmrtlib/linux/CMakeFiles/igfxcmrt.dir/__/agnostic/hardware/cm_queue.cpp.o
cc1plus: sorry, unimplemented: 64-bit mode not compiled in
make[2]: *** [media_driver/CMakeFiles/iHD_drv_video_OBJ.dir/build.make:76: media_driver/CMakeFiles/iHD_drv_video_OBJ.dir/agnostic/common/cm/cm_array.cpp.o] Error 1
make[2]: *** Waiting for unfinished jobs....
[  1%] Building CXX object cmrtlib/linux/CMakeFiles/igfxcmrt.dir/__/agnostic/hardware/cm_perf_statistics.cpp.o
cc1plus: sorry, unimplemented: 64-bit mode not compiled in
make[2]: *** [media_driver/CMakeFiles/iHD_drv_video_OBJ.dir/build.make:90: media_driver/CMakeFiles/iHD_drv_video_OBJ.dir/agnostic/common/cm/cm_buffer_rt.cpp.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:221: media_driver/CMakeFiles/iHD_drv_video_OBJ.dir/all] Error 2
[  1%] Building CXX object cmrtlib/linux/CMakeFiles/igfxcmrt.dir/__/agnostic/hardware/cm_kernel_debugger.cpp.o
[  1%] Building CXX object cmrtlib/linux/CMakeFiles/igfxcmrt.dir/share/cm_performance.cpp.o
[  1%] Building CXX object cmrtlib/linux/CMakeFiles/igfxcmrt.dir/hardware/cm_device_os.cpp.o
[  1%] Building CXX object cmrtlib/linux/CMakeFiles/igfxcmrt.dir/hardware/cm_surface_manager_os.cpp.o
[  1%] Building CXX object cmrtlib/linux/CMakeFiles/igfxcmrt.dir/hardware/cm_device_export_os.cpp.o
[  1%] Building CXX object cmrtlib/linux/CMakeFiles/igfxcmrt.dir/hardware/cm_timer_os.cpp.o
[  2%] Linking CXX shared library libigfxcmrt.so
[  2%] Built target igfxcmrt
make: *** [Makefile:156: all] Error 2
builder for '/nix/store/jynx74kc1n7q0wfsmzk3wzsrkzsc0w93-intel-media-driver-21.3.3.drv' failed with exit code 2
cannot build derivation '/nix/store/5631yxw34z6phkqkrh34mhb0c4clgwp8-shell.drv': 1 dependencies couldn't be built
error: build of '/nix/store/5631yxw34z6phkqkrh34mhb0c4clgwp8-shell.drv' failed
````

### Additional context

Found with https://github.com/guibou/nixGL/issues/81


### Notify maintainers

@primeos @jfrankenau 

### Metadata
Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

```console
nix-shell -p nix-info --run "nix-info -m"
warning: unknown setting 'experimental-features'
warning: unknown setting 'experimental-features'
 - system: `"x86_64-linux"`
 - host os: `Linux 5.14.7, NixOS, 21.11 (Porcupine)`
 - multi-user?: `yes`
 - sandbox: `yes`
warning: unknown setting 'experimental-features'
 - version: `nix-env (Nix) 2.3.15`
warning: unknown setting 'experimental-features'
 - channels(root): `"nixos-21.11pre319562.c21ba4f7bb4"`
warning: unknown setting 'experimental-features'
 - channels(guillaume): `"home-manager"`
warning: unknown setting 'experimental-features'
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`
 ```

Maintainer information:
```yaml
# a list of nixpkgs attributes affected by the problem
attribute:
# a list of nixos modules affected by the problem
module:
```

--- a/pkgs/development/libraries/intel-gmmlib/default.nix
+++ b/pkgs/development/libraries/intel-gmmlib/default.nix
@@ -11,7 +11,7 @@ stdenv.mkDerivation rec {
   src = fetchFromGitHub {
     owner = "intel";
     repo = "gmmlib";
-    rev = "${pname}-${version}";
+    rev = "intel-gmmlib-${version}";
     sha256 = "0dzqfgbd0fxl8rxgf5nmj1jd4izzaqfb0s53l96qwz1j57q5ybj5";
   };
 
@@ -27,6 +27,6 @@ stdenv.mkDerivation rec {
       OpenCL(TM) and the Intel(R) Media Driver for VAAPI.
     '';
     platforms = [ "x86_64-linux" "i686-linux" ];
-    maintainers = with maintainers; [ primeos ];
+    maintainers = with maintainers; [ primeos SuperSandro2000 ];
   };
 }

--- a/pkgs/development/libraries/intel-media-driver/default.nix
+++ b/pkgs/development/libraries/intel-media-driver/default.nix
@@ -25,6 +25,7 @@ stdenv.mkDerivation rec {
   };
 
   patches = [
+    # fix platform detection
     (fetchpatch {
       url = "https://salsa.debian.org/multimedia-team/intel-media-driver-non-free/-/raw/master/debian/patches/0002-Remove-settings-based-on-ARCH.patch";
       sha256 = "sha256-f4M0CPtAVf5l2ZwfgTaoPw7sPuAP/Uxhm5JSHEGhKT0=";

--- a/pkgs/development/libraries/intel-media-driver/default.nix
+++ b/pkgs/development/libraries/intel-media-driver/default.nix
@@ -1,7 +1,14 @@
-{ lib, stdenv, fetchFromGitHub
-, cmake, pkg-config
-, libva, libpciaccess, intel-gmmlib
-, enableX11 ? stdenv.isLinux, libX11
+{ lib
+, stdenv
+, fetchFromGitHub
+, fetchpatch
+, cmake
+, pkg-config
+, libva
+, libpciaccess
+, intel-gmmlib
+, enableX11 ? stdenv.isLinux
+, libX11
 }:
 
 stdenv.mkDerivation rec {
@@ -11,11 +18,18 @@ stdenv.mkDerivation rec {
   outputs = [ "out" "dev" ];
 
   src = fetchFromGitHub {
-    owner  = "intel";
-    repo   = "media-driver";
-    rev    = "intel-media-${version}";
+    owner = "intel";
+    repo = "media-driver";
+    rev = "intel-media-${version}";
     sha256 = "1ch1bvqg6p0i7ahblhy0h9c43y2mfhqb25v1s344iqsrywwcpzzr";
   };
+
+  patches = [
+    (fetchpatch {
+      url = "https://salsa.debian.org/multimedia-team/intel-media-driver-non-free/-/raw/master/debian/patches/0002-Remove-settings-based-on-ARCH.patch";
+      sha256 = "sha256-f4M0CPtAVf5l2ZwfgTaoPw7sPuAP/Uxhm5JSHEGhKT0=";
+    })
+  ];
 
   cmakeFlags = [
     "-DINSTALL_DRIVER_SYSCONF=OFF"
@@ -23,6 +37,8 @@ stdenv.mkDerivation rec {
     # Works only on hosts with suitable CPUs.
     "-DMEDIA_RUN_TEST_SUITE=OFF"
   ];
+
+  NIX_CFLAGS_COMPILE = lib.optionalString (stdenv.hostPlatform.system == "i686-linux") "-D_FILE_OFFSET_BITS=64";
 
   nativeBuildInputs = [ cmake pkg-config ];
 
@@ -45,6 +61,6 @@ stdenv.mkDerivation rec {
     changelog = "https://github.com/intel/media-driver/releases/tag/intel-media-${version}";
     license = with licenses; [ bsd3 mit ];
     platforms = platforms.linux;
-    maintainers = with maintainers; [ primeos jfrankenau ];
+    maintainers = with maintainers; [ primeos jfrankenau SuperSandro2000 ];
   };
 }

